File: backend/utils/sentry.py
"""
Sentry Error Tracking Integration

Provides error tracking, performance monitoring, and user context for debugging.
"""

import os
import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from sentry_sdk.integrations.logging import LoggingIntegration
import logging
import structlog

logger = structlog.get_logger(__name__)


def init_sentry() -> bool:
    """
    Initialize Sentry SDK if DSN is configured.

    Returns True if Sentry was initialized, False otherwise.

    Environment variables:
    - SENTRY_DSN: Sentry Data Source Name (required)
    - SENTRY_ENVIRONMENT: Environment name (production, staging, development)
    - SENTRY_RELEASE: Release/version identifier
    - SENTRY_TRACES_SAMPLE_RATE: Performance monitoring sample rate (0.0-1.0)
    - SENTRY_PROFILES_SAMPLE_RATE: Profiling sample rate (0.0-1.0)
    """
    dsn = os.getenv("SENTRY_DSN")

    if not dsn:
        logger.info("sentry_disabled", reason="SENTRY_DSN not configured")
        return False

    environment = os.getenv("SENTRY_ENVIRONMENT", "development")
    release = os.getenv("SENTRY_RELEASE", "travelmind@1.0.0")
    traces_sample_rate = float(os.getenv("SENTRY_TRACES_SAMPLE_RATE", "0.1"))
    profiles_sample_rate = float(os.getenv("SENTRY_PROFILES_SAMPLE_RATE", "0.1"))

    sentry_sdk.init(
        dsn=dsn,
        environment=environment,
        release=release,

        # Performance Monitoring
        traces_sample_rate=traces_sample_rate,
        profiles_sample_rate=profiles_sample_rate,

        # Integrations
        integrations=[
            FastApiIntegration(
                transaction_style="endpoint"
            ),
            SqlalchemyIntegration(),
            LoggingIntegration(
                level=logging.INFO,
                event_level=logging.ERROR
            ),
        ],

        # Data scrubbing - remove sensitive data
        send_default_pii=False,

        # Before send hook for additional filtering
        before_send=before_send_filter,

        # Ignore certain errors
        ignore_errors=[
            KeyboardInterrupt,
            SystemExit,
        ],
    )

    logger.info(
        "sentry_initialized",
        environment=environment,
        release=release,
        traces_sample_rate=traces_sample_rate
    )

    return True


def before_send_filter(event, hint):
    """
    Filter events before sending to Sentry.

    - Remove sensitive data
    - Filter out non-critical errors
    - Add additional context
    """
    # Don't send 4xx client errors (they're not bugs)
    if "exception" in event:
        exc_info = hint.get("exc_info")
        if exc_info:
            exc_type, exc_value, _ = exc_info

            # Filter out expected HTTP exceptions
            from fastapi import HTTPException
            if isinstance(exc_value, HTTPException):
                if 400 <= exc_value.status_code < 500:
                    return None  # Don't send client errors

    # Scrub sensitive headers
    if "request" in event and "headers" in event["request"]:
        headers = event["request"]["headers"]
        sensitive_headers = ["authorization", "cookie", "x-api-key"]
        for header in sensitive_headers:
            if header in headers:
                headers[header] = "[Filtered]"

    return event


def set_user_context(user_id: int, username: str, email: str = None):
    """
    Set user context for Sentry events.

    Call this after successful authentication to attach user info to errors.
    """
    sentry_sdk.set_user({
        "id": str(user_id),
        "username": username,
        "email": email,
    })


def clear_user_context():
    """Clear user context (call on logout)."""
    sentry_sdk.set_user(None)


def capture_message(message: str, level: str = "info", **extra):
    """
    Capture a message to Sentry.

    Useful for tracking non-exception events.
    """
    with sentry_sdk.push_scope() as scope:
        for key, value in extra.items():
            scope.set_extra(key, value)
        sentry_sdk.capture_message(message, level=level)


def capture_exception(exception: Exception, **extra):
    """
    Capture an exception to Sentry with extra context.
    """
    with sentry_sdk.push_scope() as scope:
        for key, value in extra.items():
            scope.set_extra(key, value)
        sentry_sdk.capture_exception(exception)


def add_breadcrumb(message: str, category: str = "custom", level: str = "info", **data):
    """
    Add a breadcrumb for debugging context.

    Breadcrumbs are logged events that lead up to an error.
    """
    sentry_sdk.add_breadcrumb(
        message=message,
        category=category,
        level=level,
        data=data
    )

File: frontend/src/utils/sentry.js
/**
 * Sentry Error Tracking Integration
 *
 * Provides error tracking and performance monitoring for the frontend.
 */

import * as Sentry from '@sentry/react'

/**
 * Initialize Sentry SDK
 *
 * Environment variables (via Vite):
 * - VITE_SENTRY_DSN: Sentry Data Source Name
 * - VITE_SENTRY_ENVIRONMENT: Environment name
 * - VITE_SENTRY_RELEASE: Release version
 */
export function initSentry() {
  const dsn = import.meta.env.VITE_SENTRY_DSN

  if (!dsn) {
    console.log('Sentry: DSN not configured, error tracking disabled')
    return false
  }

  const environment = import.meta.env.VITE_SENTRY_ENVIRONMENT || 'development'
  const release = import.meta.env.VITE_SENTRY_RELEASE || 'travelmind-frontend@1.0.0'

  Sentry.init({
    dsn,
    environment,
    release,

    // Performance Monitoring
    tracesSampleRate: import.meta.env.PROD ? 0.1 : 1.0,

    // Session Replay (optional, captures user sessions)
    replaysSessionSampleRate: 0.1,
    replaysOnErrorSampleRate: 1.0,

    // Integrations
    integrations: [
      Sentry.browserTracingIntegration(),
      // Mask all text and inputs in session replay so secrets the user types
      // (AI API keys, passwords — incl. when the key field is toggled to plain
      // text) are never sent to Sentry.
      Sentry.replayIntegration({
        maskAllText: true,
        maskAllInputs: true,
        blockAllMedia: true,
      }),
    ],

    // Filter out non-critical errors
    beforeSend(event, hint) {
      const error = hint.originalException

      // Ignore network errors (user offline, etc.)
      if (error?.message?.includes('Network Error')) {
        return null
      }

      // Ignore cancelled requests
      if (error?.code === 'ECONNABORTED') {
        return null
      }

      // Ignore 401/403 errors (expected auth errors)
      if (error?.response?.status === 401 || error?.response?.status === 403) {
        return null
      }

      return event
    },

    // Don't send PII by default
    sendDefaultPii: false,
  })

  console.log(`Sentry: Initialized (${environment})`)
  return true
}

/**
 * Set user context for error tracking
 * Call after successful login
 */
export function setUserContext(user) {
  if (!user) return

  Sentry.setUser({
    id: String(user.id),
    username: user.username,
    email: user.email,
  })
}

/**
 * Clear user context
 * Call on logout
 */
export function clearUserContext() {
  Sentry.setUser(null)
}

/**
 * Capture an exception manually
 */
export function captureException(error, context = {}) {
  Sentry.withScope((scope) => {
    Object.entries(context).forEach(([key, value]) => {
      scope.setExtra(key, value)
    })
    Sentry.captureException(error)
  })
}

/**
 * Capture a message
 */
export function captureMessage(message, level = 'info', context = {}) {
  Sentry.withScope((scope) => {
    Object.entries(context).forEach(([key, value]) => {
      scope.setExtra(key, value)
    })
    Sentry.captureMessage(message, level)
  })
}

/**
 * Add a breadcrumb for debugging
 */
export function addBreadcrumb(message, category = 'custom', data = {}) {
  Sentry.addBreadcrumb({
    message,
    category,
    data,
    level: 'info',
  })
}

/**
 * Sentry Error Boundary wrapper component
 */
export const SentryErrorBoundary = Sentry.ErrorBoundary

/**
 * Sentry Profiler for performance monitoring
 */
export const SentryProfiler = Sentry.withProfiler

/**
 * Einen gefangenen Fehler melden, statt ihn wegzuwerfen.
 *
 * Am 2026-08-25 fanden sich sieben `catch (err) { toast.error(…) }`-Bloecke, in
 * denen der Fehler selbst verworfen wurde: Der Nutzer sah eine allgemeine
 * Meldung, und es gab keinerlei Spur, woran es lag — weder in der Konsole noch
 * in Sentry. eslint hat sie nur als "unbenutzte Variable" gemeldet.
 *
 * Der Toast bleibt die Nachricht an den Nutzer; diese Funktion ist die Spur
 * fuer die Fehlersuche.
 *
 * @param {unknown} fehler   Das gefangene Objekt.
 * @param {string}  kontext  Wo es passiert ist, z.B. 'AdminPanel.loadData'.
 * @param {object}  extra    Zusaetzliche Angaben (keine personenbezogenen Daten).
 */
export function reportError(fehler, kontext, extra = {}) {
  // Immer sichtbar in der Browser-Konsole — auch ohne konfiguriertes Sentry.
  console.error(`[${kontext}]`, fehler)

  Sentry.captureException(fehler, {
    tags: { kontext },
    extra,
  })
}
